[mlir][emitc] Add a `declare_func` operation

marbre opened the pull request:
This adds the `emitc.declare_func` operation that allows to emit the declaration of an `emitc.func` at a specific location.

marbre committed:
[mlir][emitc] Add a `declare_func` operation

This adds the `emitc.declare_func` operation that allows to emit the
declaration of an `emitc.func` at a specific location.

marbre commented:
@AlexDenisov if you don't mind I would like to ask you for a code review.

simon-camp reviewed (changes requested):
I just took a quick look and will review in more detail tomorrow.

simon-camp commented on mlir/test/Target/Cpp/declare_func.mlir:
```diff
@@ -0,0 +1,16 @@
+// CHECK: int32_t bar(int32_t [[V1:[^ ]*]]);
+emitc.declare_func @bar
+// CHECK: int32_t bar(int32_t [[V1:[^ ]*]]) {
+emitc.func @bar(%arg0: i32) -> i32 {
```
Can you change this to call @foo below, like in the op description. This shows the usefulness/need of the op better.

marbre replied:
So far most of our test "just" test the "core" functionality, whereas transporting the usefulness is left to the description. I agree that this would make a more meaningful example and we should provide examples somewhere. However, till now I was not sure if should use the tests for this. With this thoughts in mind, WDYT?

simon-camp replied:
Sure, we can test that in a potential future legalization pass.

AlexDenisov commented on mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:
````diff
@@ -460,6 +460,48 @@ def EmitC_CallOp : EmitC_Op<"call",
+    }
+    ```
+  }];
+  let arguments = (ins FlatSymbolRefAttr:$callee);
````
FWIW, it's often referred to as `callee` in other dialects, though there is no unified name.

simon-camp replied:
I've checked some upstream dialects (async, emitc, func, llvm) and I only found that symbol being named callee from, well, call-like operations. :)

marbre replied:
Yeah, I need to correct my previous statement as i was wrong and the corresponding attribute is named `sym_name` in `emitc.func`. I will adopt this for the `FlatSymbolRefAttr`of the `emitc.declare_func`op.

AlexDenisov commented on mlir/lib/Dialect/EmitC/IR/EmitC.cpp:
```diff
@@ -393,6 +393,24 @@ FunctionType CallOp::getCalleeType() {
+  // Check that the callee attribute was specified.
+  auto fnAttr = (*this)->getAttrOfType<FlatSymbolRefAttr>("callee");
+  if (!fnAttr)
+    return emitOpError("requires a 'callee' symbol reference attribute");
```
It's better to use `getCalleeAttrName()` around these lines to ensure that the messages are consistent.

marbre replied:
I had the same in mind but more or less followed what `func.call`does (we do the same for `emitc.call`):

https://github.com/llvm/llvm-project/blob/6a3fde6d600cccd2ffbede6dd54519036cc4089c/mlir/lib/Dialect/Func/IR/FuncOps.cpp#L64-L66

But I am happy to revise this and adjust it for `emitc.call` as well.

marbre replied:
(Will adjust `emitc.call` in a follow up if this is landed this way.)

AlexDenisov reviewed (commented):
That's a great and much needed feature. I used some terrible workarounds to achieve the same, and ended up using `VerbatimOp`.

I'd add a test case for when the prototype is emitted inside the function body as well.

It looks like this feature would be incomplete without support of external functions (`mlir::func::FuncOp` which are only declaring the functions).
I'd expect [this case](https://github.com/llvm/llvm-project/blob/e7d40a87ff230528131541f6ac17a2e1a7dc78e1/mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp#L60-L62) to naturally map to this operation? Otherwise the use case is rather limited.

marbre commented:
> That's a great and much needed feature. I used some terrible workarounds to achieve the same, and ended up using `VerbatimOp`.
> 
> I'd add a test case for when the prototype is emitted inside the function body as well.

Well @simon-camp and me already started discussion the value of a function declaration within a function body. While this is legal in C, it is not as soon as it is a static function. Therefore, I see three option:

* Don't allow function declarations in function bodies at all.
* Allow function declarations in function bodies, but disallow declarations for static functions.
* Allow function declarations anyway and check if the output results in legal C in a validation pass (a validation pass is something I am currently working on).

> It looks like this feature would be incomplete without support of external functions (`mlir::func::FuncOp` which are only declaring the functions). I'd expect [this case](https://github.com/llvm/llvm-project/blob/e7d40a87ff230528131541f6ac17a2e1a7dc78e1/mlir/lib/Conversion/FuncToEmitC/FuncToEmitC.cpp#L60-L62) to naturally map to this operation? Otherwise the use case is rather limited.

To make sure I understand what you're proposing: A `func.func` with empty body should be converted to a `emitc.declare_func` with an `extern` specifier?

AlexDenisov commented:
> Allow function declarations in function bodies, but disallow declarations for static functions.
> Allow function declarations anyway and check if the output results in legal C in a validation pass (a validation pass is something I am currently working on).

These two sound better than disallowing them completely.

> To make sure I understand what you're proposing: A func.func with empty body should be converted to a emitc.declare_func with an extern specifier?

Yes, exactly. Although, I'm not sure about the `extern` specifier: it would help to have a mechanism to specify `extern "C"` without having a conflict.

marbre commented:
> I agree with the need to declare external functions. We can't rewrite an external func.func to a emitc.declare_func though. The former provides a symbol while the latter refrences an existing symbol. We can convert it to an emitc.func with empty body and adjust the emitter to print it correctly as a declaration. Then `extern "C"` would simply go into the specifier attribute.

I'll prototype something and see if we have other options. If those don't work out and if we follow this approach, we could (more or less) land the current implementation of `emitc.declare_func` as is. Any objections @simon-camp @AlexDenisov?

marbre committed:
Rename `callee` to `sym_name` and revise checking if specified

marbre commented:
> None other than resolving the two open inline discussions. 👍

For now I added a verifier that checks if the parent is a `ModuleOp`. However, I am not sure if this is too restrictive. E.g. what about function declarations within IREE's VM modules @simon-camp? Is this still something we want or need to do? If though, I would rather drop to model this restriction as part of the IR and would rather go for a validator.

simon-camp commented on mlir/lib/Dialect/EmitC/IR/EmitC.cpp:
```diff
@@ -393,6 +393,36 @@ FunctionType CallOp::getCalleeType() {
+  auto funcOp =
+      SymbolTable::lookupNearestSymbolFrom<emitc::FuncOp>((*this), fnAttr);
+
+  if (!isa<ModuleOp>((*this)->getParentOp()) && funcOp.isPrivate())
```
We shouldn't be using `isPrivate` here. The symbol visibility is not neccessarily in sync with the specifiers.

marbre replied:
If it's not, something other is broken but I agree that can be the case. That's another point why a validation pass might be more favorable here. Beside that, going through all specifiers args in the array adds more a much larger overhead to the verifier.

marbre commented:
> > For now I added a verifier that checks if the parent is a `ModuleOp`. However, I am not sure if this is too restrictive. E.g. what about function declarations within IREE's VM modules @simon-camp? Is this still something we want or need to do? If though, I would rather drop to model this restriction as part of the IR and would rather go for a validator.
> 
> I would leave this out for now and see how it works out. Seems easier to add restrictions later.

I dropped the change and will add it to our todo list for the validation pass.
